### PR TITLE
CM-1069: sessions and users

### PR DESCRIPTION
**Pull request by cyplas**
* fixes bug in assigning quiz to session (not user)
* more generally, adds assign_session function and use it in routes if and only if the route has jwt_token and session_uuid (though not sure if this is the optimal criterion?)

### CODE REVIEW

**Review by danmash — CHANGES_REQUESTED** [review]
@cyplas thank you for a fast PR, I really appreciate this. 
Let me clarify something in my following message in slack:

> or I would say login/register or any other request with a JWT token should update session’s user_uuid field

I more like solutions which allow us to cover "any other request". I would prefer solutions # 2,3,4 from [this answer](https://stackoverflow.com/a/51820573/2207154) (not decorator) . You are free to use any solution you like. The acceptance criteria is avoid calling `assign_session` function directly from any endpoint. Having this we could don't worry about this bug in the future. All requests will be covered.

Hope this makes sense.

**Comment by danmash** [comment]
@cyplas Let me know if you are still working on this

**Comment by cyplas** [comment]
> I more like solutions which allow us to cover "any other request". I would prefer solutions # 2,3,4 from [this answer](https://stackoverflow.com/a/51820573/2207154) (not decorator) . You are free to use any solution you like. The acceptance criteria is avoid calling `assign_session` function directly from any endpoint. Having this we could don't worry about this bug in the future. All requests will be covered.

@danmash Good idea. I opted for solution 2. Now let's see if I understood you right. :) And I'm not sure if placing the hook in climatemind.py is appropriate, but I'm not sure where would be better?

**Comment by danmash** [comment]
@cyplas thank you for the update.
Let me think about your question.
Meanwhile, I suggest you add the following unittests because I don't know how to test these changes in a different way. 

1. [Create session factory without user](https://github.com/ClimateMind/climatemind-backend/blob/34772d090c8665cbf4199a54e35673d1a15dac8d/app/account/tests/test_account_routes.py#L276-L277)
2. [Make a request without access token](https://github.com/ClimateMind/climatemind-backend/blob/34772d090c8665cbf4199a54e35673d1a15dac8d/app/account/tests/test_account_routes.py#L287-L291) and make sure user is not populated
3. Then [create a user + access token and run the same request](https://github.com/ClimateMind/climatemind-backend/blob/34772d090c8665cbf4199a54e35673d1a15dac8d/app/account/tests/test_account_routes.py#L36-L42) with cookie added. Assert the session user field populated successfully. 
4. [parameterize](https://github.com/ClimateMind/climatemind-backend/blob/f3bff070928ab941b271d5ce7fc6c3753edf0f2d/app/feed/tests/test_process_alignment_feed.py#L65) the test with multiple different URLs to make sure `before_request` works fine for any request

**Comment by cyplas** [comment]
> I suggest you add the following unittests because I don't know how to test these changes in a different way.
> 
> 1. [Create session factory without user](https://github.com/ClimateMind/climatemind-backend/blob/34772d090c8665cbf4199a54e35673d1a15dac8d/app/account/tests/test_account_routes.py#L276-L277)
> 2. [Make a request without access token](https://github.com/ClimateMind/climatemind-backend/blob/34772d090c8665cbf4199a54e35673d1a15dac8d/app/account/tests/test_account_routes.py#L287-L291) and make sure user the session's user field is not populated
> 3. Then [create a user + access token and run the same request](https://github.com/ClimateMind/climatemind-backend/blob/34772d090c8665cbf4199a54e35673d1a15dac8d/app/account/tests/test_account_routes.py#L36-L42) with cookie added. Assert the session user field populated successfully.
> 4. [parameterize](https://github.com/ClimateMind/climatemind-backend/blob/f3bff070928ab941b271d5ce7fc6c3753edf0f2d/app/feed/tests/test_process_alignment_feed.py#L65) the test with multiple different URLs to make sure `before_request` works fine for any request

I've tried to do something along these lines now, but I haven't written these kinds of tests before, so I'm fumbling my way along and clearly not there yet. The final assert fails, even though doing an analogous real API call works, so I think it's my test that's the problem, not the function. @danmash Your help would be appreciated.

**Review by danmash — CHANGES_REQUESTED** [review]

**Comment by danmash** [comment]
@cyplas Great progress! I think the problem is the wrong argument order, see my suggestions above.

File: app/session/tests/test_session_helpers.py [hunk 1]
```diff
@@ -59,3 +63,42 @@ def test_get_ip_address_prod(mocked_check_if_local):
...
+    assert response.status_code == route_data["code"], "Expected response status."
+    assert (
+        session.user_uuid is None
+    ), "No user should be assigned to session yet (session={}, user={}).".format(
+        session.session_uuid, session.user_uuid
+    )
+
+    user = UsersFactory()
+    access_token = create_access_token(identity=user, fresh=True)
+    auth_header = [("Authorization", "Bearer " + access_token)]
```
[the remark this answers is not in the document]
Comment (cyplas) [reply app/session/tests/test_session_helpers.py:100]: > I'm not sure auth header will work, if not use[ cookies instead](https://github.com/ClimateMind/climatemind-backend/blob/47460cece9de2f0af46396bcf110bf0410cd05b4/app/conftest.py#L47-L46)

It seems to be working. And this is the way it is being called by the front-end, as far as I know.

File: app/session/tests/test_session_helpers.py [hunk 2]
```diff
@@ -59,3 +63,42 @@ def test_get_ip_address_prod(mocked_check_if_local):
 
     assert get_ip_address(request) == first_ip
     assert get_ip_address(request) == first_ip
+
+
+@pytest.mark.parametrize(
+    "route_data", [{"url": "myths.get_general_myths", "json": {}, "code": 200}]
+)
+def test_maybe_assign_session(client, accept_json, route_data):
```
[the remark this answers is not in the document]
Comment (cyplas) [reply app/session/tests/test_session_helpers.py]: > arguments from parametrize should be before fixtures

ok, done

File: app/session/tests/test_session_helpers.py [hunk 3]
```diff
@@ -59,3 +63,42 @@ def test_get_ip_address_prod(mocked_check_if_local):
...
+
+    session = SessionsFactory(user=None)
+    session_header = [("X-Session-Id", session.session_uuid)]
+
+    response = client.get(
+        url_for(route_data["url"]),
```
[the remark this answers is not in the document]
Comment (cyplas) [reply app/session/tests/test_session_helpers.py]: > replace `route_data` with a specific variable `url`

I refactored the parametrised arguments.

**Comment by cyplas** [comment]
> I think the problem is the wrong argument order, see my suggestions above.

The main problem turned out to be that the @before_request was only being added to the app in climatemind.py, which is separate from the conftest.py one.

**Review by danmash — DISMISSED** [review]
Perfect! now we could be sure the bug will not appear in any new endpoints as well.

**Pull request merged**